### PR TITLE
feat(web): add surround selection in composer

### PR DESCRIPTION
**Pull request by mbuvarp**
# What

Add surround-selection support in the composer so that when text is selected and the user types a supported surround symbol, the selection is wrapped instead of replaced.

This also includes dead-key backtick support, so layouts that produce `` ` `` through composition can still surround the selected text.

Currently applies to these symbols: `(`, `[`, `{`, `'`, `"`, `` ` ``, `<`, `«`, `*`, `_`

## Why

I understand this is a feature and not a bug fix, and might land outside of what you are likely to accept re: the contribution guidelines. But it's something I always miss in editors when they don't have it, since I habitually select text and press parenthesis, quote or whatever.

If this is something that should be opt-in behind a setting, I can add that as well.

## Checklist

- [x] This PR is small and focused
- [x] I explained what changed and why
- [x] I included a video for animation/interaction changes

## Video

https://github.com/user-attachments/assets/f138dfaf-8bbd-4752-852a-9a738c5dbc78

---

> [!NOTE]
> **Medium Risk**
> Intercepts `keydown`/`beforeinput`/composition events in the Lexical composer to rewrite prompt text and selection, which can regress typing/IME behavior around mentions and inline tokens despite added tests.
> 
> **Overview**
> **Composer typing enhancement:** when text is selected in the chat composer and the user types a supported opening symbol (e.g. `(`, `"`, `` ` ``, `*`), the selection is wrapped with the matching close symbol and the inner selection is preserved for repeated wrapping.
> 
> Adds a new `ComposerSurroundSelectionPlugin` that hooks `keydown`/`beforeinput`/`input`/`compositionend`, skips wrapping when the selection is collapsed, touches inline tokens, or crosses mention-boundary whitespace, and merges dead-key backtick composition into a single undo step.
> 
> Extends mention parsing utilities with `selectionTouchesMentionBoundary` and fixes `splitPromptIntoComposerSegments` to preserve consecutive terminal-context placeholders; adds browser/integration tests covering wrapping, backward selections, dead-key composition, and mention edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8daa8786cb8bae20fc73eb604ba39e90663b4d1e. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Add surround selection to the composer for wrapping selected text with paired symbols
> - Adds `ComposerSurroundSelectionPlugin` in [ComposerPromptEditor.tsx](https://github.com/pingdotgg/t3code/pull/1502/files#diff-12fcb05b3e147eba955638d8df57440b8cf7f796941af0989c9bfa57b5b4f4a4) that intercepts `beforeinput`/`keydown` events: when a recognized open symbol (e.g. `(`, `[`, `'`, `*`, `«`) is typed with a non-collapsed selection, the plugin wraps the selected text with the corresponding pair and keeps the inner text selected for successive wrapping.
> - Selections that touch mention boundary whitespace or inline tokens fall back to normal replacement behavior.
> - Adds `selectionTouchesMentionBoundary` in [composer-editor-mentions.ts](https://github.com/pingdotgg/t3code/pull/1502/files#diff-fe3b0266af75d6e06d8819ee307b1048dd163e07bfc6758027f84a441c92be3c) to detect when a selection includes mention-adjacent whitespace.
> - Dead-key backtick composition is detected and transformed into a backtick surround in a single undo step via `HISTORY_MERGE_TAG`.
> - Fixes `splitPromptIntoComposerSegments` to preserve consecutive terminal-context placeholders as distinct segments.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 8daa878.</sup>
> 
>

### CODE REVIEW

File: apps/web/src/components/ComposerPromptEditor.tsx [hunk 1]
```diff
@@ -878,6 +942,242 @@ function ComposerInlineTokenBackspacePlugin() {
...
+            const backtickCloseSymbol = BACKTICK_SURROUND_CLOSE_SYMBOL;
+            if (backtickCloseSymbol === null) {
+              pendingDeadKeySelectionRef.current = null;
+              return;
+            }
+
+            const expectedResolvedValue = `${pendingDeadKeySelection.value.slice(0, pendingDeadKeySelection.start)}\`${pendingDeadKeySelection.value.slice(pendingDeadKeySelection.end)}`;
+            if (currentValue !== expectedResolvedValue) {
+              return;
+            }
```
[the remark this answers is not in the document]
Comment (mbuvarp) [reply apps/web/src/components/ComposerPromptEditor.tsx]: Fixed.

I kept the working dead-key flow intact, but now clear the pending dead-key/surround refs on terminal failure paths:
- after `compositionend` if the editor value still does not match the expected resolved backtick replacement
- if the minimal replacement path fails to produce a `RangeSelection`

Validation:
- `bun fmt apps/web/src/components/ComposerPromptEditor.tsx`
- `bun lint apps/web/src/components/ComposerPromptEditor.tsx`
- `bun run test:browser src/components/ChatView.browser.tsx`

File: apps/web/src/components/ComposerPromptEditor.tsx [hunk 2]
```diff
@@ -1072,6 +1392,8 @@ function ComposerPromptEditorInner({
         cursor: nextCursor,
         expandedCursor: nextExpandedCursor,
         terminalContextIds,
+        selectionStart: selectionRange.start,
+        selectionEnd: selectionRange.end,
```
[the remark this answers is not in the document]
Comment (mbuvarp) [reply apps/web/src/components/ComposerPromptEditor.tsx]: Fixed.

I removed `selectionStart` / `selectionEnd` from the no-op snapshot comparison so pure selection changes no longer trigger redundant `onChange` callbacks.

Validation:
- `bun fmt apps/web/src/components/ComposerPromptEditor.tsx apps/web/src/components/ChatView.browser.tsx`
- `bun lint apps/web/src/components/ComposerPromptEditor.tsx apps/web/src/components/ChatView.browser.tsx`
- `bun run test:browser src/components/ChatView.browser.tsx`

File: apps/web/src/components/ComposerPromptEditor.tsx [hunk 3]
```diff
@@ -901,6 +1229,8 @@ function ComposerPromptEditorInner({
     cursor: initialCursor,
     expandedCursor: expandCollapsedComposerCursor(value, initialCursor),
     terminalContextIds: terminalContexts.map((context) => context.id),
+    selectionStart: initialCursor,
+    selectionEnd: initialCursor,
```
[the remark this answers is not in the document]
Comment (mbuvarp) [reply apps/web/src/components/ComposerPromptEditor.tsx]: Fixed.

I removed the unused `selectionStart` / `selectionEnd` snapshot fields and the dead helper that became unused after the earlier offset fix.

Validation:
- `bun fmt apps/web/src/components/ComposerPromptEditor.tsx`
- `bun lint apps/web/src/components/ComposerPromptEditor.tsx`
- `bun run test:browser src/components/ChatView.browser.tsx`

File: apps/web/src/components/ComposerPromptEditor.tsx [hunk 4]
```diff
@@ -878,6 +942,253 @@ function ComposerInlineTokenBackspacePlugin() {
...
+      rootElement?.addEventListener("beforeinput", onBeforeInput, true);
+      rootElement?.addEventListener("input", onInput);
+      rootElement?.addEventListener("compositionend", onCompositionEnd);
+      activeRootElement = rootElement;
+    });
+    const unregisterBeforeInputCommand = editor.registerCommand(
+      BEFORE_INPUT_COMMAND,
+      () => false,
+      COMMAND_PRIORITY_HIGH,
+    );
```
[the remark this answers is not in the document]
Comment (mbuvarp) [reply apps/web/src/components/ComposerPromptEditor.tsx]: Fixed.

I removed the no-op `BEFORE_INPUT_COMMAND` registration and its now-unused import. The native capturing `beforeinput` listener is the only path that actually handles this feature.

Validation:
- `bun fmt apps/web/src/components/ComposerPromptEditor.tsx`
- `bun lint apps/web/src/components/ComposerPromptEditor.tsx`
- `bun run test:browser src/components/ChatView.browser.tsx`

File: apps/web/src/components/ComposerPromptEditor.tsx [hunk 5]
```diff
@@ -878,6 +941,247 @@ function ComposerInlineTokenBackspacePlugin() {
...
+      }
+
+      if (event.defaultPrevented || event.isComposing || event.metaKey || event.ctrlKey) {
+        pendingSurroundSelectionRef.current = null;
+        pendingDeadKeySelectionRef.current = null;
+        return;
+      }
+
+      const shouldTrackDeadKeyBacktick =
+        BACKTICK_SURROUND_CLOSE_SYMBOL !== null && event.key === "Dead" && event.shiftKey;
```
[the remark this answers is not in the document]
Comment (mbuvarp) [reply apps/web/src/components/ComposerPromptEditor.tsx]: Fixed.

I removed the `event.shiftKey` requirement so dead-key tracking now starts from any `event.key === "Dead"` while backtick surround is enabled. The later rewrite still only applies when the composition actually resolves to a literal backtick, so the path stays layout-agnostic without broadening the final surround behavior.

Validation:
- `bun fmt apps/web/src/components/ComposerPromptEditor.tsx`
- `bun lint apps/web/src/components/ComposerPromptEditor.tsx`
- `bun run test:browser src/components/ChatView.browser.tsx`

File: apps/web/src/composer-editor-mentions.ts [hunk 6]
```diff
@@ -64,6 +68,44 @@ function splitPromptTextIntoComposerSegments(text: string): ComposerPromptSegmen
...
+    }
+
+    if (
+      afterMentionIndex < prompt.length &&
+      /\s/.test(prompt[afterMentionIndex] ?? "") &&
+      rangeIncludesIndex(start, end, afterMentionIndex)
+    ) {
+      return true;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (mbuvarp) [reply apps/web/src/composer-editor-mentions.ts]: Fixed.

I changed `selectionTouchesMentionBoundary` to scan mentions using the same placeholder-aware text slicing as the composer mention parser, so mentions immediately following `INLINE_TERMINAL_CONTEXT_PLACEHOLDER` are now detected correctly.

I also added a regression test for the `\uFFFC@AGENTS.md there` case.

Validation:
- `bun fmt apps/web/src/composer-editor-mentions.ts apps/web/src/composer-editor-mentions.test.ts apps/web/src/components/ComposerPromptEditor.tsx`
- `bun lint apps/web/src/composer-editor-mentions.ts apps/web/src/composer-editor-mentions.test.ts apps/web/src/components/ComposerPromptEditor.tsx`
- `cd apps/web && bun run test src/composer-editor-mentions.test.ts`

I also tried `cd apps/web && bun run test:browser src/components/ChatView.browser.tsx`, but it is currently blocked by the merged upstream `@t3tools/client-runtime` package-resolution issue rather than this change.

File: apps/web/src/composer-editor-mentions.ts [hunk 7]
```diff
@@ -29,6 +33,44 @@ function pushTextSegment(segments: ComposerPromptSegment[], text: string): void
...
+    }
+    textCursor = index + 1;
+  }
+
+  if (textCursor < prompt.length && visitor(prompt.slice(textCursor), textCursor) === true) {
+    return true;
+  }
+
+  return false;
+}
```
[the remark this answers is not in the document]
Comment (mbuvarp) [reply apps/web/src/composer-editor-mentions.ts:86]: Fixed.

I extracted the placeholder walk into a shared `forEachPromptSegmentSlice(...)` helper and now use that traversal in both `forEachPromptTextSlice(...)` and `splitPromptIntoComposerSegments(...)`, so the terminal-context placeholder splitting logic only lives in one place.

I also added a regression test for consecutive placeholders to make sure the refactor preserves prompt positions.

Validation:
- `cd apps/web && bun run test src/composer-editor-mentions.test.ts`
- `bun fmt`
- `bun lint`
- `bun typecheck`

**Review by juliusmarminge — APPROVED** [review]
